src/connector/src/source/filesystem/opendal_source/opendal_reader.rs

```diff
@@ -285,3 +285,85 @@ pub fn extract_valid_column_indices(
+) -> bool {
+    match rw_data_type {
+        risingwave_common::types::DataType::Boolean => {
+            matches!(arrow_data_type, arrow_schema_iceberg::DataType::Boolean)
+        }
+        risingwave_common::types::DataType::Int16 => {
+            matches!(arrow_data_type, arrow_schema_iceberg::DataType::Int16)
+        }
+        risingwave_common::types::DataType::Int32 => {
+            matches!(arrow_data_type, arrow_schema_iceberg::DataType::Int32)
+        }
+        risingwave_common::types::DataType::Int64 => {
+            matches!(arrow_data_type, arrow_schema_iceberg::DataType::Int64)
+        }
+        risingwave_common::types::DataType::Float32 => {
+            matches!(arrow_data_type, arrow_schema_iceberg::DataType::Float32)
+        }
+        risingwave_common::types::DataType::Float64 => {
+            matches!(arrow_data_type, arrow_schema_iceberg::DataType::Float64)
+        }
+        risingwave_common::types::DataType::Decimal => {
+            matches!(
+                arrow_data_type,
+                arrow_schema_iceberg::DataType::Decimal128(_, _)
+            ) || matches!(
+                arrow_data_type,
+                arrow_schema_iceberg::DataType::Decimal256(_, _)
+            )
+        }
+        risingwave_common::types::DataType::Date => {
+            matches!(arrow_data_type, arrow_schema_iceberg::DataType::Date32)
+                || matches!(arrow_data_type, arrow_schema_iceberg::DataType::Date64)
+        }
+        risingwave_common::types::DataType::Varchar => {
+            matches!(arrow_data_type, arrow_schema_iceberg::DataType::Utf8)
+        }
+        risingwave_common::types::DataType::Time => {
+            matches!(arrow_data_type, arrow_schema_iceberg::DataType::Time32(_))
+                || matches!(arrow_data_type, arrow_schema_iceberg::DataType::Time64(_))
+        }
+        risingwave_common::types::DataType::Timestamp => {
+            matches!(
+                arrow_data_type,
+                arrow_schema_iceberg::DataType::Timestamp(_, _)
+            )
+        }
+        risingwave_common::types::DataType::Timestamptz => {
+            matches!(
+                arrow_data_type,
+                arrow_schema_iceberg::DataType::Timestamp(_, _)
+            )
+        }
+        risingwave_common::types::DataType::Interval => {
+            matches!(arrow_data_type, arrow_schema_iceberg::DataType::Interval(_))
+        }
+        risingwave_common::types::DataType::List(inner_type) => {
+            if let arrow_schema_iceberg::DataType::List(field_ref) = arrow_data_type {
+                let inner_rw_type = inner_type.clone();
+                let inner_arrow_type = field_ref.data_type();
+                is_data_type_matching(&inner_rw_type, inner_arrow_type)
+            } else {
+                false
+            }
+        }
+        risingwave_common::types::DataType::Map(_) => {
+            // Directly return false for Map types
```

Comment:
I think `Map` type in RW and arrow is quite different, the former is `pub struct MapType(Box<(DataType, DataType)>);` while the latter is `Arc<Field>`, so I just reject this type comparison. Pls correct me if I was wrong.

Reply:
cc @xxchan

Reply:
No, they have the same representation. Arrow's `Arc<Field>` is just a struct with 2 fields. 

Please check `src/common/src/array/arrow/arrow_impl.rs` for the conversion code.